flink/v1.15/flink/src/main/java/org/apache/iceberg/flink/sink/committer/FilesCommitter.java

```diff
@@ -0,0 +1,314 @@
+  private static final String FLINK_JOB_ID = "flink.job-id";
+  private static final int MAX_RECOMMIT_TIMES = 3;
+
+  private static final String MAX_COMMITTED_CHECKPOINT_ID = "flink.max-committed-checkpoint-id";
+
+  // TableLoader to load iceberg table lazily.
+  private final TableLoader tableLoader;
+  private final boolean replacePartitions;
+  private final Map<String, String> snapshotProperties;
+
+  // It will have an unique identifier for one job.
+  private final Table table;
+  private transient Map<String, Long> maxCommittedCheckpointIdForJob = Maps.newHashMap();
+
+  private transient ExecutorService workerPool;
+
+  public FilesCommitter(
+      TableLoader tableLoader,
+      boolean replacePartitions,
+      Map<String, String> snapshotProperties,
+      int workerPoolSize) {
+    this.tableLoader = tableLoader;
+    this.replacePartitions = replacePartitions;
+    this.snapshotProperties = snapshotProperties;
+
+    this.workerPool =
+        ThreadPools.newWorkerPool(
+            "iceberg-worker-pool-" + Thread.currentThread().getName(), workerPoolSize);
+
+    // Open the table loader and load the table.
+    this.tableLoader.open();
+    this.table = tableLoader.loadTable();
+  }
+
+  @Override
+  public void commit(Collection<CommitRequest<FilesCommittable>> requests)
+      throws IOException, InterruptedException {
+    int dataFilesNumStored = 0;
+    int deleteFilesNumStored = 0;
+
+    Collection<CommitRequest<FilesCommittable>> store = Lists.newArrayList();
+    String jobId = null;
+    long maxCommittableCheckpointId = INITIAL_CHECKPOINT_ID;
+
+    for (CommitRequest<FilesCommittable> request : requests) {
+      WriteResult committable = request.getCommittable().committable();
+      Long committableCheckpointId = request.getCommittable().checkpointId();
+      jobId = request.getCommittable().jobID();
+      if (request.getNumberOfRetries() > MAX_RECOMMIT_TIMES) {
+        String message =
+            String.format(
+                "Failed to commit transaction %s after retrying %d times",
+                request.getCommittable(), MAX_RECOMMIT_TIMES);
+        request.signalFailedWithUnknownReason(new RuntimeException(message));
+      } else if (request.getNumberOfRetries() > 0) {
+        commitResult(
+            committable.dataFiles().length,
+            committable.deleteFiles().length,
+            Lists.newArrayList(request));
+      } else if (committableCheckpointId
+              > maxCommittedCheckpointIdForJob.getOrDefault(jobId, Long.MAX_VALUE)
+          || getMaxCommittedCheckpointId(table, jobId) == -1) {
+        store.add(request);
+        dataFilesNumStored = dataFilesNumStored + committable.dataFiles().length;
+        deleteFilesNumStored = deleteFilesNumStored + committable.deleteFiles().length;
+        maxCommittableCheckpointId = committableCheckpointId;
+      } else if (committableCheckpointId > getMaxCommittedCheckpointId(table, jobId)) {
```

Comment:
Why do we commit these separately?

Reply:
The reason for this is the same as committing data for `request.getNumberOfRetries() > 0` separately. `committableCheckpointId > getMaxCommittedCheckpointId(table, jobId)` means data is restore from the previous job, i will try to commit them first.